### PR TITLE
DL4J + Keras import: Causal Conv1D support

### LINKED ISSUE #7629: Libnd4j: Add temporal convolution support (causal conv1d op)
See section 3.2 here: https://arxiv.org/pdf/1803.01271.pdf

Idea is very simple: it's basically standard conv1d with 2 small changes:
* Amount of padding so that output size is same as input size
* Causal convolutions

Causal 1d convolutions are standard 1d convolutions but shifted so that their receptive field is based on current and past time steps, not future time steps. Same math, just offset receptive field.

image

### LINKED ISSUE #7968: DL4J Keras Import: Import TCN (causal conv1d) keras model to DL4J
Hi,

I am trying to import TCN model implemented with keras-tcn https://github.com/philipperemy/keras-tcn library 

While importing : 

`    val modelKerasTcn = KerasModelImport.importKerasModelAndWeights("data/model_tcn.h5", false)
`

I get an error: 

`Exception in thread "main" org.deeplearning4j.nn.modelimport.keras.exceptions.UnsupportedKerasConfigurationException: Unsupported convolution border mode: causal. Please file an issue at http://github.com/deeplearning4j/deeplearning4j/issues.
	at org.deeplearning4j.nn.modelimport.keras.layers.convolutional.KerasConvolutionUtils.getConvolutionModeFromConfig(KerasConvolutionUtils.java:269)
	at org.deeplearning4j.nn.modelimport.keras.layers.convolutional.KerasConvolution1D.<init>(KerasConvolution1D.java:107)
	at org.deeplearning4j.nn.modelimport.keras.utils.KerasLayerUtils.getKerasLayerFromConfig(KerasLayerUtils.java:233)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.prepareLayers(KerasModel.java:218)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.<init>(KerasModel.java:164)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.<init>(KerasModel.java:96)
	at org.deeplearning4j.nn.modelimport.keras.utils.KerasModelBuilder.buildModel(KerasModelBuilder.java:307)
	at org.deeplearning4j.nn.modelimport.keras.KerasModelImport.importKerasModelAndWeights(KerasModelImport.java:155)`

#### Version Information
DL4J : Version 1.0.0-beta4

Is it something you consider to include in your future release? 

Thank you,
Sandra

### PR DESCRIPTION
**Pull request by AlexDBlack**
DL4J causal conv1d support, and Keras causal conv1d support.
DL4J part is essentially done, remaining tasks:
* keras import tests not yet passing
* Validation for not using ConvolutionMode.Causal for conv2d/3d
* Javadoc for ConvolutionMode

Fixes: https://github.com/eclipse/deeplearning4j/issues/7968
Fixes: https://github.com/eclipse/deeplearning4j/issues/7629

Also fixes some keras import bugs for NCW/channels_first format.
Considerably improves the conv1d import tests.
Also makes nIn inference for RNNs more robust for NCW/NWC cases.
https://github.com/KonduitAI/dl4j-test-resources/pull/11

### CODE REVIEW

**Comment by AlexDBlack** [comment]
Closed in favor of: https://github.com/KonduitAI/deeplearning4j/pull/107
